gtsfm/utils/verification.py

```diff
@@ -42,32 +42,62 @@ def recover_relative_pose_from_essential_matrix(
 
 
-def create_essential_matrix(i2Ri1: Rot3, i2Ui1: Unit3) -> EssentialMatrix:
-    """Creates essential matrix from rot3 and unit3.
+def fundamental_to_essential_matrix(
+    i2Fi1: np.ndarray, camera_intrinsics_i1: Cal3Bundler, camera_intrinsics_i2: Cal3Bundler
+) -> np.ndarray:
+    """Converts the fundamental matrix to essential matrix using camera intrinsics.
 
     Args:
-        i2Ri1: relative rotation.
-        i2Ui1: relative translation direction.
+        i2Fi1: fundamental matrix which maps points in image #i1 to lines in image #i2.
+        camera_intrinsics_i1: intrinsics for image #i1.
+        camera_intrinsics_i2: intrinsics for image #i2.
 
     Returns:
-        essential matrix i2Ei1.
+        Estimated essential matrix i2Ei1 as numpy array of shape (3x3).
     """
-    return EssentialMatrix(i2Ri1, i2Ui1)
+    return camera_intrinsics_i2.K().T @ i2Fi1 @ camera_intrinsics_i1.K()
 
 
-def fundamental_to_essential_matrix(
-    i2Fi1: np.ndarray,
-    camera_intrinsics_i1: Cal3Bundler,
-    camera_intrinsics_i2: Cal3Bundler,
+def essential_to_fundamental_matrix(
+    i2Ei1: EssentialMatrix, camera_intrinsics_i1: Cal3Bundler, camera_intrinsics_i2: Cal3Bundler
 ) -> np.ndarray:
-    """Converts the fundamental matrix to essential matrix using camera intrinsics.
+    """Converts the essential matrix to fundamental matrix using camera intrinsics.
 
     Args:
-        i2Fi1: fundamental matrix which maps points in image #i1 to lines in image #i2.
+        i2Ei1: essential matrix which maps points in image #i1 to lines in image #i2.
         camera_intrinsics_i1: intrinsics for image #i1.
         camera_intrinsics_i2: intrinsics for image #i2.
 
     Returns:
-            Estimated essential matrix i2Ei1 as numpy array of shape (3x3).
+        Fundamental matrix i2Fi1 as numpy array of shape (3x3).
     """
-    return camera_intrinsics_i2.K().T @ i2Fi1 @ camera_intrinsics_i1.K()
+    return np.linalg.inv(camera_intrinsics_i2.K().T) @ i2Ei1.matrix() @ np.linalg.inv(camera_intrinsics_i1.K())
+
+
+def compute_epipolar_distances(
+    normalized_coords_i1: np.ndarray, normalized_coords_i2: np.ndarray, i2Ei1: EssentialMatrix
+) -> np.ndarray:
+    """Compute symmetric point-line epipolar distances between normalized coordinates of correspondences.
+
+    Args:
+        normalized_coords_i1: normalized coordinates in image i1, of shape Nx2.
+        normalized_coords_i2: corr. normalized coordinates in image i2, of shape Nx2.
+        i2Ei1: essential matrix between two images
+
+    Returns:
+        Symmetric epipolar distances for each row of the input, of shape N.
+    """
+    # construct the essential matrix in the opposite directin
+    i2Ti1 = Pose3(i2Ei1.rotation(), i2Ei1.direction().point3())
+    i1Ti2 = i2Ti1.inverse()
+    i1Ei2 = EssentialMatrix(i1Ti2.rotation(), Unit3(i1Ti2.translation()))
+
+    # get lines in i2 and i1
+    epipolar_lines_i2 = feature_utils.convert_to_epipolar_lines(normalized_coords_i1, i2Ei1)
```

Comment:
@ayushbaid we need to check if these are None right? if so, let's return NaN?

Reply:
NaN is more appropriate for elements in the array, right? How can I return a NaN in place of a numpy array?

Reply:
forgot we were returning an array of distances.

i guess "None" then?

Reply:
Yes None is better imo.